File: modules/expo-mpv-player/ios/MpvPlayerView.swift
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * Inspired by and/or derived from Streamyfin (https://github.com/streamyfin/streamyfin)
 * and Findroid (https://github.com/findroid/findroid).
 * Copyright (c) the original authors and Seanime Tenji contributors.
 */

import ExpoModulesCore
import AVKit
import AVFoundation
import CoreMedia
import MPVKit
import UIKit

struct NowPlayingMetadata {
    let title: String?
    let artist: String?
    let albumTitle: String?
    let artworkUri: String?
}

/// Configuration passed from JSto load a video
struct VideoLoadConfig {
    let url: URL
    let headers: [String: String]?
    let externalSubtitles: [(url: String, title: String?)]
    let startPosition: Double?
    let autoplay: Bool
}

/// The native ExpoView that wraps an AVSampleBufferDisplayLayer and MPVLayerRenderer.
/// All async functions defined in the Expo module's View { } block are dispatched to this class.
final class MpvSurfaceExpoView: ExpoView, MPVLayerRendererDelegate, PiPControllerDelegate {

    // MARK: - Properties

    private let displayLayer = AVSampleBufferDisplayLayer()
    private var renderer: MPVLayerRenderer?
    private var pipController: PiPController?
    private var hasStartedRenderer = false
    private var isZoomedFill = false
    private let nowPlayingManager = MPVNowPlayingManager.shared
    private var lastNowPlayingSyncAt: CFAbsoluteTime = 0
    private var lastLayoutSize: CGSize = .zero

    // Pending config waiting for renderer to start
    private var pendingConfig: VideoLoadConfig?

    // Event emitters set by the Expo module
    let onLoad = EventDispatcher()
    let onProgress = EventDispatcher()
    let onPlaybackStateChange = EventDispatcher()
    let onError = EventDispatcher()
    let onTracksReady = EventDispatcher()

    // MARK: - Init

    required init(appContext: AppContext? = nil) {
        super.init(appContext: appContext)

        // Set up the display layer
        displayLayer.videoGravity = .resizeAspect
        displayLayer.preventsDisplaySleepDuringVideoPlayback = true

        layer.addSublayer(displayLayer)

        // Audio session
        configureAudioSession()

#if !os(tvOS)
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(onPhoneLock),
            name: UIApplication.protectedDataWillBecomeUnavailableNotification,
            object: nil
        )
#endif
    }

    override func layoutSubviews() {
        super.layoutSubviews()
        displayLayer.frame = bounds

        guard bounds.size != lastLayoutSize else { return }
        lastLayoutSize = bounds.size

        if pipController?.isPictureInPictureActive ?? false {
            applyPictureInPictureVideoGravity()
        } else {
            applyInlineVideoGravity()
        }
        renderer?.refreshVideoPresentation()
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
        clearNowPlayingInfo()
        renderer?.stop()
        renderer = nil
    }

    @objc private func onPhoneLock() {
        pause()
    }

    // MARK: - Audio Session

    private func configureAudioSession() {
        let session = AVAudioSession.sharedInstance()
        do {
            try session.setCategory(.playback, mode: .moviePlayback)
            try session.setActive(true)
        } catch {
            print("[MpvPlayerView] Audio session error: \(error)")
        }
    }

    // MARK: - Source Prop

    func setSource(_ config: VideoLoadConfig) {
        if !hasStartedRenderer {
            startRenderer()
        }
        guard renderer != nil else {
            pendingConfig = config
            return
        }
        load(config)
    }

    private func startRenderer() {
        let newRenderer = MPVLayerRenderer(displayLayer: displayLayer)
        newRenderer.delegate = self
        do {
            try newRenderer.start()
        } catch {
            print("[MpvPlayerView] Failed to start renderer: \(error)")
            onError(["error": "\(error)"])
            return
        }
        renderer = newRenderer
        hasStartedRenderer = true

        // Set up PiP
        pipController = PiPController(sampleBufferDisplayLayer: displayLayer)
        pipController?.delegate = self

        // Play pending config if any
        if let config = pendingConfig {
            pendingConfig = nil
            load(config)
        }
    }

    private func load(_ config: VideoLoadConfig) {
        renderer?.load(
            url: config.url,
            headers: config.headers,
            externalSubtitles: config.externalSubtitles,
            startPosition: config.startPosition
        )

        if config.autoplay {
            play()
        }

        onLoad(["url": config.url.absoluteString])
    }

    private func applyInlineVideoGravity() {
        displayLayer.videoGravity = isZoomedFill ? .resizeAspectFill : .resizeAspect
    }

    private func applyPictureInPictureVideoGravity() {
        displayLayer.videoGravity = .resizeAspect
    }

    func setSpeed(_ speed: Double) { renderer?.setSpeed(speed) }
    func getSpeed() -> Double { return renderer?.getSpeed() ?? 1.0 }

    func isPaused() -> Bool { return renderer?.isPausedState ?? true }
    func getCurrentPosition() -> Double { return renderer?.cachedPosition ?? 0 }
    func getDuration() -> Double { return renderer?.cachedDuration ?? 0 }

    // MARK: - PiP Controls

    func startPictureInPicture() {
        renderer?.setPictureInPictureRenderingModeEnabled(true)
        applyPictureInPictureVideoGravity()

        if pipController?.startPictureInPicture() != true {
            renderer?.setPictureInPictureRenderingModeEnabled(false)
            applyInlineVideoGravity()
        }
    }

    func stopPictureInPicture() { pipController?.stopPictureInPicture() }
    func isPictureInPictureSupported() -> Bool { return pipController?.isPictureInPictureSupported ?? false }
    func isPictureInPictureActive() -> Bool { return pipController?.isPictureInPictureActive ?? false }

    // MARK: - Now Playing

    func setNowPlayingMetadata(_ metadata: NowPlayingMetadata?) {
        guard let metadata else {
            nowPlayingManager.clear()
            return
        }

        nowPlayingManager.setMetadata(
            title: metadata.title,
            artist: metadata.artist,
            albumTitle: metadata.albumTitle,
            artworkUrl: metadata.artworkUri
        )
        syncNowPlaying(force: true)
    }

    // MARK: - Subtitle Controls

    func getSubtitleTracks() -> [[String: Any]] { return renderer?.getSubtitleTracks() ?? [] }
    func getChapters() -> [[String: Any]] { return renderer?.getChapters() ?? [] }
    func setSubtitleTrack(_ trackId: Int) { renderer?.setSubtitleTrack(trackId) }
    func disableSubtitles() { renderer?.disableSubtitles() }
    func getCurrentSubtitleTrack() -> Int { return renderer?.getCurrentSubtitleTrack() ?? 0 }
    func addSubtitleFile(_ url: String, select: Bool) { renderer?.addSubtitleFile(url: url, select: select) }
    func setSubtitleDelay(_ delay: Double) { renderer?.setSubtitleDelay(delay) }
    func setSubtitleFontSize(_ size: Int) { renderer?.setSubtitleFontSize(size) }
    func setSubtitleVisibility(_ visible: Bool) { renderer?.setSubtitleVisibility(visible) }
    func setSubtitlePosition(_ position: Int) { renderer?.setSubtitlePosition(position) }
    func setSubtitleScale(_ scale: Double) { renderer?.setSubtitleScale(scale) }
    func setSubtitleMarginY(_ margin: Int) { renderer?.setSubtitleMarginY(margin) }
    func setSubtitleAlignX(_ alignment: String) { renderer?.setSubtitleAlignX(alignment) }
    func setSubtitleAlignY(_ alignment: String) { renderer?.setSubtitleAlignY(alignment) }

    // MARK: - Audio Controls

    func getAudioTracks() -> [[String: Any]] { return renderer?.getAudioTracks() ?? [] }
    func setAudioTrack(_ trackId: Int) { renderer?.setAudioTrack(trackId) }
    func getCurrentAudioTrack() -> Int { return renderer?.getCurrentAudioTrack() ?? 0 }
    func setAudioDelay(_ delay: Double) { renderer?.setAudioDelay(delay) }

    // MARK: - Zoom

    func setVideoZoom(_ scale: Double) {
        isZoomedFill = scale > 1.001
        if !(pipController?.isPictureInPictureActive ?? false) {
            applyInlineVideoGravity()
        }
        renderer?.setVideoZoom(scale)
    }

    func setZoomedToFill(_ zoomed: Bool) {
        isZoomedFill = zoomed
        if !(pipController?.isPictureInPictureActive ?? false) {
            applyInlineVideoGravity()
        }
    }

    func isZoomedToFill() -> Bool {
        return isZoomedFill
    }

    // MARK: - Technical Info

    func getTechnicalInfo() -> [String: Any] {
        return renderer?.getTechnicalInfo() ?? [:]
    }

    // MARK: - Now Playing Helpers

    private func setupRemoteCommands() {
        nowPlayingManager.setupRemoteCommands(
            playHandler: { [weak self] in self?.play() },
            pauseHandler: { [weak self] in self?.pause() },
            toggleHandler: { [weak self] in
                guard let self else { return }
                if self.renderer?.isPausedState ?? true {
                    self.play()
                } else {
                    self.pause()
                }
            },
            seekHandler: { [weak self] position in self?.seekTo(position) },
            skipForward: { [weak self] interval in self?.seekBy(interval) },
            skipBackward: { [weak self] interval in self?.seekBy(-interval) }
        )
    }

    private func currentPlaybackRate() -> Float {
        guard let renderer, !renderer.isPausedState else { return 0 }
        return Float(renderer.playbackSpeed)
    }

    private func syncNowPlaying(force: Bool = false) {
        let now = CFAbsoluteTimeGetCurrent()
        if !force && (now - lastNowPlayingSyncAt) < 1 {
            return
        }
        lastNowPlayingSyncAt = now
        nowPlayingManager.updatePlayback(
            position: renderer?.cachedPosition ?? 0,
            duration: renderer?.cachedDuration ?? 0,
            playbackRate: currentPlaybackRate()
        )
    }

    private func clearNowPlayingInfo() {
        nowPlayingManager.cleanupRemoteCommands()
        nowPlayingManager.clear()
        nowPlayingManager.deactivateAudioSession()
    }

    // MARK: - MPVLayerRendererDelegate

    func renderer(_ renderer: MPVLayerRenderer, didUpdatePosition position: Double, duration: Double, cacheSeconds: Double) {
        syncNowPlaying()
        onProgress([
            "position": position,
            "duration": duration,
            "cacheSeconds": cacheSeconds,
        ])

        // Update PiP timebase
        pipController?.setCurrentTimeFromSeconds(position, duration: duration)
    }

    func renderer(_ renderer: MPVLayerRenderer, didChangePause isPaused: Bool) {
        syncNowPlaying(force: true)
        onPlaybackStateChange([
            "isPaused": isPaused,
            "isPlaying": !isPaused,
        ])

        // Update PiP rate
        pipController?.setPlaybackRate(isPaused ? 0 : Float(renderer.playbackSpeed))
        pipController?.updatePlaybackState()
    }

    func renderer(_ renderer: MPVLayerRenderer, didChangeLoading isLoading: Bool) {
        onPlaybackStateChange(["isLoading": isLoading])
    }

    func renderer(_ renderer: MPVLayerRenderer, didBecomeReadyToSeek: Bool) {
        onPlaybackStateChange(["isReadyToSeek": true])
    }

    func renderer(_ renderer: MPVLayerRenderer, didBecomeTracksReady: Bool) {
        onTracksReady([:])
    }

    func renderer(_ renderer: MPVLayerRenderer, didChangeEOF eofReached: Bool) {
        onPlaybackStateChange(["eofReached": eofReached])
    }

    func renderer(_ renderer: MPVLayerRenderer, didChangeSpeed speed: Double) {
        onPlaybackStateChange(["speed": speed])
        if !renderer.isPausedState {
            pipController?.setPlaybackRate(Float(speed))
        }
        syncNowPlaying(force: true)
    }

    func renderer(_ renderer: MPVLayerRenderer, didChangeSubtitleDelay delay: Double) {
        onPlaybackStateChange(["subtitleDelay": delay])
    }

    func renderer(_ renderer: MPVLayerRenderer, didChangeAudioDelay delay: Double) {
        onPlaybackStateChange(["audioDelay": delay])
    }

    // MARK: - PiPControllerDelegate

    func pipController(_ controller: PiPController, willStartPictureInPicture: Bool) {
        renderer?.setPictureInPictureRenderingModeEnabled(true)
        applyPictureInPictureVideoGravity()
        onPlaybackStateChange(["isPiPActive": true])
    }

    func pipController(_ controller: PiPController, didStartPictureInPicture: Bool) {
        guard !didStartPictureInPicture else { return }
        renderer?.setPictureInPictureRenderingModeEnabled(false)
        applyInlineVideoGravity()
        onPlaybackStateChange(["isPiPActive": false])
    }

    func pipController(_ controller: PiPController, willStopPictureInPicture: Bool) {}

    func pipController(_ controller: PiPController, didStopPictureInPicture: Bool) {
        renderer?.setPictureInPictureRenderingModeEnabled(false)
        applyInlineVideoGravity()
        onPlaybackStateChange(["isPiPActive": false])
    }

    func pipController(_ controller: PiPController, didTransitionToRenderSize renderSize: CMVideoDimensions) {
        applyPictureInPictureVideoGravity()
        renderer?.refreshVideoPresentation()
    }

    func pipController(_ controller: PiPController, restoreUserInterfaceForPictureInPictureStop completionHandler: @escaping (Bool) -> Void) {
        completionHandler(true)
    }

    func pipControllerPlay(_ controller: PiPController) {
        play()
    }

    func pipControllerPause(_ controller: PiPController) {
        pause()
    }

    func pipController(_ controller: PiPController, skipByInterval interval: CMTime) {
        let seconds = CMTimeGetSeconds(interval)
        seekBy(seconds)
    }

    func pipControllerIsPlaying(_ controller: PiPController) -> Bool {
        return !(renderer?.isPausedState ?? true)
    }

    func pipControllerDuration(_ controller: PiPController) -> Double {
        return renderer?.cachedDuration ?? 0
    }

    func pipControllerCurrentPosition(_ controller: PiPController) -> Double {
        return renderer?.cachedPosition ?? 0
    }

    // MARK: - Playback Controls

    func play() {
        nowPlayingManager.activateAudioSession()
        setupRemoteCommands()
        renderer?.play()
        syncNowPlaying(force: true)
    }

    func pause() {
        renderer?.pause()
        syncNowPlaying(force: true)
    }

    func seekTo(_ position: Double) {
        renderer?.seek(to: position)
        syncNowPlaying(force: true)
    }

    func seekBy(_ offset: Double) {
        renderer?.seek(by: offset)
        syncNowPlaying(force: true)
    }
}

File: modules/expo-mpv-player/android/src/main/java/expo/modules/mpvplayer/MpvPlayerView.kt
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * Inspired by and/or derived from Streamyfin (https://github.com/streamyfin/streamyfin)
 * and Findroid (https://github.com/findroid/findroid).
 * Copyright (c) the original authors and Seanime Tenji contributors.
 */

package expo.modules.mpvplayer

import android.content.Context
import android.graphics.Color
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.SurfaceHolder
import android.view.SurfaceView
import android.view.ViewGroup
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.viewevent.EventDispatcher
import expo.modules.kotlin.views.ExpoView

private const val TAG = "MpvPlayerView"

data class VideoLoadConfig(
    val sourceId: String? = null,
    val url: String,
    val headers: Map<String, String>? = null,
    val externalSubtitles: List<Pair<String, String?>>? = null,
    val startPosition: Double? = null,
    val autoplay: Boolean = true
)

class MpvPlayerView(context: Context, appContext: AppContext) : ExpoView(context, appContext),
    MPVLayerRenderer.Delegate, SurfaceHolder.Callback, PiPController.Delegate {

    val onLoad by EventDispatcher()
    val onPlaybackStateChange by EventDispatcher()
    val onProgress by EventDispatcher()
    val onError by EventDispatcher()
    val onTracksReady by EventDispatcher()
    val onPictureInPictureChange by EventDispatcher()

    private var surfaceView: SurfaceView
    private var renderer: MPVLayerRenderer? = null
    private var pipController: PiPController? = null

    private var currentUrl: String? = null
    private var currentSourceId: String? = null
    private var cachedPosition: Double = 0.0
    private var cachedDuration: Double = 0.0
    private var intendedPlayState: Boolean = false
    private var surfaceReady: Boolean = false
    private var pendingConfig: VideoLoadConfig? = null
    private var pendingVideoOutput: VideoOutput? = null
    private var canApplyPendingSource: Boolean = false
    private var _isZoomedToFill: Boolean = false
    private var dispatchedPiPActive: Boolean = false
    private var dispatchedPaused: Boolean? = null
    private var hostForeground: Boolean = true
    private var pipRecoveryPending: Boolean = false

    private var rendererStarted: Boolean = false

    private val pipHandler = Handler(Looper.getMainLooper())
    private val recoverRunnable = Runnable { recoverVideoOutput() }

    init {
        setBackgroundColor(Color.BLACK)

        surfaceView = SurfaceView(context).apply {
            layoutParams = ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT
            )
        }
        surfaceView.holder.addCallback(this)
        surfaceView.addOnLayoutChangeListener { _, left, top, right, bottom,
                                                oldLeft, oldTop, oldRight, oldBottom ->
            val width = right - left
            val height = bottom - top
            val oldWidth = oldRight - oldLeft
            val oldHeight = oldBottom - oldTop
            if (width > 0 && height > 0 && (width != oldWidth || height != oldHeight)) {
                renderer?.updateSurfaceSize(width, height)
            }
        }
        addView(surfaceView)

        renderer = MPVLayerRenderer(context).also {
            it.delegate = this
        }

        pipController = PiPController(context, appContext).also {
            it.setPlayerView(surfaceView)
            it.delegate = this
        }
    }

    override fun surfaceCreated(holder: SurfaceHolder) {
        surfaceReady = true

        if (rendererStarted) {
            renderer?.attachSurface(holder.surface)
            syncSurfaceSize()
        }

        applyPendingSourceR()

        if (hostForeground && pipRecoveryPending) {
            scheduleRecovery()
        }
    }

    override fun surfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int) {
        if (width > 0 && height > 0) {
            renderer?.updateSurfaceSize(width, height)
        }
    }

    override fun surfaceDestroyed(holder: SurfaceHolder) {
        surfaceReady = false
        renderer?.detachSurface()
    }

    private fun syncSurfaceSize() {
        if (!surfaceReady) return
        val width = surfaceView.width
        val height = surfaceView.height
        if (width > 0 && height > 0) {
            renderer?.updateSurfaceSize(width, height)
        }
    }

    fun stageVideoOutput(value: String) {
        val output = VideoOutput.fromMpvValue(value)
        if (output == null) {
            Log.w(TAG, "Ignoring unsupported video output: $value")
            return
        }
        pendingVideoOutput = output
    }

    fun stageVideo(config: VideoLoadConfig) {
        if (config.url == currentUrl && config.sourceId == currentSourceId) return
        if (config.url == pendingConfig?.url && config.sourceId == pendingConfig?.sourceId) return
        pendingConfig = config
        canApplyPendingSource = false
    }

    fun applyPendingProps() {
        pendingVideoOutput?.let { output ->
            pendingVideoOutput = null
            renderer?.setVideoOutput(output)
        }

        if (!rendererStarted) {
            try {
                renderer?.start()
                rendererStarted = true
            } catch (error: Throwable) {
                val message = describeError(error)
                Log.e(TAG, "Could not initialize native player", error)
                onError(mapOf("error" to message))
                canApplyPendingSource = false
                return
            }
            surfaceView.holder.surface?.takeIf { surfaceReady && it.isValid }?.let { surface ->
                renderer?.attachSurface(surface)
                syncSurfaceSize()
            }
        }

        canApplyPendingSource = true
        applyPendingSourceR()
    }

    private fun describeError(error: Throwable): String {
        val parts = mutableListOf<String>()
        val seen = mutableSetOf<Throwable>()
        var current: Throwable? = error

        while (current != null && seen.add(current)) {
            val detail = current.message?.trim()
            val label = current::class.java.simpleName
            parts += if (detail.isNullOrEmpty()) label else "$label: $detail"
            current = current.cause
        }

        return parts.joinToString("; ").take(1200).ifBlank {
            "Unknown player initialization error"
        }
    }

    private fun applyPendingSourceR() {
        if (!surfaceReady || !canApplyPendingSource) return
        val config = pendingConfig ?: return
        pendingConfig = null
        loadVideoInternal(config)
    }

    private fun loadVideoInternal(config: VideoLoadConfig) {
        currentUrl = config.url
        currentSourceId = config.sourceId
        cachedPosition = 0.0
        cachedDuration = 0.0
        dispatchedPaused = null

        renderer?.load(
            url = config.url,
            headers = config.headers,
            startPosition = config.startPosition,
            externalSubtitles = config.externalSubtitles
        )

        if (config.autoplay) {
            play()
        }

        onLoad(mapOf("url" to config.url))
    }

    fun play() {
        intendedPlayState = true
        renderer?.play()
        pipController?.setPlaybackRate(1.0)
        dispatchPauseState(false)
    }

    fun pause() {
        intendedPlayState = false
        renderer?.pause()
        pipController?.setPlaybackRate(0.0)
        dispatchPauseState(true)
    }

    fun seekTo(position: Double) {
        cachedPosition = position
        renderer?.seekTo(position)
    }

    fun seekBy(offset: Double) {
        renderer?.seekBy(offset)
    }

    fun setSpeed(speed: Double) {
        renderer?.setSpeed(speed)
    }

    fun getSpeed(): Double {
        return renderer?.getSpeed() ?: 1.0
    }

    fun isPaused(): Boolean {
        return renderer?.isPaused ?: true
    }

    fun getCurrentPosition(): Double {
        return renderer?.cachedPosition ?: cachedPosition
    }

    fun getDuration(): Double {
        return renderer?.cachedDuration ?: cachedDuration
    }

    fun startPictureInPicture(): Boolean {
        val started = pipController?.startPictureInPicture() != null
        dispatchPictureInPictureState(started || isPictureInPictureActive())

        pipHandler.removeCallbacksAndMessages(null)
        pipHandler.postDelayed({ syncSurfaceSize() }, 100)
        pipHandler.postDelayed({ syncSurfaceSize() }, 500)
        return started
    }

    fun stopPictureInPicture() {
        pipHandler.removeCallbacksAndMessages(null)
        pipController?.stopPictureInPicture()
    }

    fun isPictureInPictureSupported(): Boolean {
        return pipController?.isPictureInPictureSupported() ?: false
    }

    fun isPictureInPictureActive(): Boolean {
        return pipController?.isPictureInPictureActive() ?: false
    }

    fun dispatchPictureInPictureState(active: Boolean) {
        if (dispatchedPiPActive == active) return
        dispatchedPiPActive = active
        onPlaybackStateChange(mapOf("isPiPActive" to active))
    }

    fun getSubtitleTracks(): List<Map<String, Any>> {
        return renderer?.getSubtitleTracks() ?: emptyList()
    }

    fun getChapters(): List<Map<String, Any>> {
        return renderer?.getChapters() ?: emptyList()
    }

    fun setSubtitleTrack(trackId: Int) {
        renderer?.setSubtitleTrack(trackId)
    }

    fun disableSubtitles() {
        renderer?.disableSubtitles()
    }

    fun getCurrentSubtitleTrack(): Int {
        return renderer?.getCurrentSubtitleTrack() ?: -1
    }

    fun addSubtitleFile(url: String, select: Boolean) {
        renderer?.addSubtitleFile(url, select)
    }

    fun setSubtitleDelay(delay: Double) {
        renderer?.setSubtitleDelay(delay)
    }

    fun setSubtitleFontSize(size: Int) {
        renderer?.setSubtitleFontSize(size)
    }

    fun setSubtitleVisibility(visible: Boolean) {
        renderer?.setSubtitleVisibility(visible)
    }

    fun setSubtitlePosition(position: Int) {
        renderer?.setSubtitlePosition(position)
    }

    fun getAudioTracks(): List<Map<String, Any>> {
        return renderer?.getAudioTracks() ?: emptyList()
    }

    fun setAudioTrack(trackId: Int) {
        renderer?.setAudioTrack(trackId)
    }

    fun getCurrentAudioTrack(): Int {
        return renderer?.getCurrentAudioTrack() ?: -1
    }

    fun setAudioDelay(delay: Double) {
        renderer?.setAudioDelay(delay)
    }

    fun setVideoZoom(scale: Double) {
        renderer?.setVideoZoom(scale)
    }

    fun setZoomedToFill(zoomed: Boolean) {
        _isZoomedToFill = zoomed
        renderer?.setZoomedToFill(zoomed)
    }

    fun isZoomedToFill(): Boolean {
        return _isZoomedToFill
    }

    fun getTechnicalInfo(): Map<String, Any> {
        return renderer?.getTechnicalInfo() ?: emptyMap()
    }

    fun getPlayerView(): android.view.View = surfaceView

    override fun onPositionChanged(position: Double, duration: Double, cacheSeconds: Double) {
        cachedPosition = position
        cachedDuration = duration

        if (pipController?.isPictureInPictureActive() == true) {
            pipController?.setCurrentTime(position, duration)
        }

        onProgress(
            mapOf(
                "position" to position,
                "duration" to duration,
                "cacheSeconds" to cacheSeconds
            )
        )
    }

    override fun onPauseChanged(isPaused: Boolean) {
        pipController?.setPlaybackRate(if (isPaused) 0.0 else 1.0)
        dispatchPauseState(isPaused)
    }

    private fun dispatchPauseState(isPaused: Boolean) {
        if (dispatchedPaused == isPaused) return
        dispatchedPaused = isPaused
        onPlaybackStateChange(
            mapOf(
                "isPaused" to isPaused,
                "isPlaying" to !isPaused
            )
        )
    }

    override fun onLoadingChanged(isLoading: Boolean) {
        onPlaybackStateChange(
            mapOf(
                "isLoading" to isLoading
            )
        )
    }

    override fun onReadyToSeek() {
        onPlaybackStateChange(
            mapOf(
                "isReadyToSeek" to true
            )
        )
    }

    override fun onTracksReady() {
        onTracksReady(emptyMap<String, Any>())
    }

    override fun onVideoDimensionsChanged(width: Int, height: Int) {
        pipController?.setVideoDimensions(width, height)
    }

    override fun onEOFChanged(eofReached: Boolean) {
        onPlaybackStateChange(
            mapOf(
                "eofReached" to eofReached
            )
        )
    }

    override fun onSpeedChanged(speed: Double) {
        onPlaybackStateChange(
            mapOf(
                "speed" to speed
            )
        )
    }

    override fun onSubtitleDelayChanged(delay: Double) {
        onPlaybackStateChange(
            mapOf(
                "subtitleDelay" to delay
            )
        )
    }

    override fun onAudioDelayChanged(delay: Double) {
        onPlaybackStateChange(
            mapOf(
                "audioDelay" to delay
            )
        )
    }

    override fun onError(message: String) {
        onError(
            mapOf(
                "error" to message
            )
        )
    }

    override fun onPlay() {
        play()
    }

    override fun onPause() {
        pause()
    }

    override fun onSeekBy(seconds: Double) {
        seekBy(seconds)
    }

    override fun onPictureInPictureModeChanged(isInPiP: Boolean) {
        pipHandler.removeCallbacksAndMessages(null)
        pipHandler.postDelayed({ syncSurfaceSize() }, 100)
        if (isInPiP) {
            pipRecoveryPending = false
            pipHandler.postDelayed({ syncSurfaceSize() }, 500)
        } else {
            pipRecoveryPending = true
            if (hostForeground) scheduleRecovery()
        }
        onPictureInPictureChange(mapOf("isActive" to isInPiP))
        dispatchPictureInPictureState(isInPiP)
    }

    fun setHostForeground(foreground: Boolean) {
        hostForeground = foreground
    }

    fun onHostResume() {
        hostForeground = true
        if (!rendererStarted || currentUrl == null) return
        if (isPictureInPictureActive()) return

        if (!pipRecoveryPending && (renderer?.isTv != true || intendedPlayState)) return

        scheduleRecovery()
    }

    private fun scheduleRecovery() {
        pipHandler.removeCallbacks(recoverRunnable)
        pipHandler.postDelayed(recoverRunnable, 300)
    }

    private fun recoverVideoOutput() {
        if (!hostForeground || !rendererStarted || currentUrl == null) return
        if (!surfaceReady || isPictureInPictureActive()) return

        val pipRecovery = pipRecoveryPending
        if (!pipRecovery && (renderer?.isTv != true || intendedPlayState)) return

        val surface = surfaceView.holder.surface.takeIf { it.isValid } ?: return
        val recovered = renderer?.recoverVideoOutput(
            surface = surface,
            playWhenReady = pipRecovery && intendedPlayState,
        ) == true

        if (pipRecovery && recovered) {
            pipRecoveryPending = false
        }
    }

    fun cleanup() {
        pipHandler.removeCallbacksAndMessages(null)
        pipController?.stopPictureInPicture()
        renderer?.stop()
        hostForeground = false
        pipRecoveryPending = false
        surfaceReady = false
        currentUrl = null
        currentSourceId = null
        rendererStarted = false
        renderer = null
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        cleanup()
    }
}
